### PR TITLE
Contact rollups extract step

### PR DESCRIPTION
**Pull request by bencodeorg**
Draft PR for extract step of new contact rollups. This PR contains:
- a migration for a new raw_contacts table
- duplicates a minimal version of a couple of the files used in the contact rollups process in order to allow for the first step to run
- comprehensive testing (read: a single test) for the new raw_contacts model

If you execute bin/cron/build_contact_rollups_v2 it'll take all `EmailPreferences` and extract them into `RawContacts`.

### CODE REVIEW

File: bin/cron/build_contact_rollups_v2 [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env ruby
+
+require_relative('../../dashboard/config/environment')
+require 'cdo/contact_rollups_v2'
+require 'cdo/contact_rollups_validation'
+require 'cdo/pardot'
+require 'cdo/chat_client'
+require 'cdo/only_one'
+require 'cdo/log_collector'
+require 'honeybadger/ruby'
```
Comment (bencodeorg) [inline bin/cron/build_contact_rollups_v2:10]: These are mostly extraneous at this point.

**Comment by hacodeorg** [comment]
I have a few thoughts on how we should structure this. I'd like to see your views also.

**1. Where do files live?**
- Pipeline code will be in `bin/cron/build_contact_rollups_v2.rb`. This is the glue connecting main components (data extraction, processing, diff-ing, syncing downstream and syncing upstream).
- Component (production) codes will be in `lib/cdo/contact_rollups/v2/` folder. To keep it simple, we can have 2 files, `contact_rollups.rb` and `pardot.rb` just like v1. (We can create a v1 folder to keep v1 code).
  - `contact_rollups.rb` will host code for 3 components: data extraction, data processing and data diffing.
  - `pardot.rb` will host all-Pardot related code.
- Component test codes will be in `lib/test/cdo/contact_rollups/v2/`
- DB migration code will be in `dashboard/db/migrate/`
- Models code (if any) will be in `dashboard/app/models/contact_rollups/`. 

**2. What parts of ActiveRecord we want to use?** 
My thought right now is we use ActiveRecord to make the code easier to read. I don't see any specific features of ActiveRecord that are critical to the performance of the pipeline. Specific questions are:
- Do we need a model for each table? Should we just use plain-old-ruby?
- Is it better to use `ActiveRecord::Base.connection.execute` to execute SQL command, especially for bulk insert, instead of processing row-by-row?

**Comment by bencodeorg** [comment]
I think your description of where files live make total sense. I am in favor of using ActiveRecord (and avoiding raw SQL commands) from a testability / readability perspective -- I think we can probably avoid some of the row-by-row processing by using activerecord-import:

https://github.com/zdennis/activerecord-import

File: lib/cdo/contact_rollups_v2.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+require 'cdo/properties'
+require 'json'
+
+class ContactRollupsV2
+  def self.build_contact_rollups(log_collector)
+    # Set opt_in based on information collected in Dashboard Email Preference.
+    update_email_preferences
+  end
+
+  def self.update_email_preferences
```
Comment (hacodeorg) [inline lib/cdo/contact_rollups_v2.rb]: how about `extract_email_preferences`? We will have more extraction functions later.

File: lib/cdo/contact_rollups_v2.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+require 'state_abbr'
+require src_dir 'database'
+require_relative('../../dashboard/config/environment')
+require 'cdo/properties'
+require 'json'
+
+class ContactRollupsV2
+  def self.build_contact_rollups(log_collector)
+    # Set opt_in based on information collected in Dashboard Email Preference.
+    update_email_preferences
```
Comment (hacodeorg) [inline lib/cdo/contact_rollups_v2.rb]: We need to truncate the raw_contacts table before running data extraction

Comment (hacodeorg) [reply lib/cdo/contact_rollups_v2.rb]: You can time this function by using `log_collector.time!("<function_name>") {<function>}`

File: bin/cron/build_contact_rollups_v2 [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+CHAT_CHANNEL = "cron-daily".freeze
+
+def main
+  log_collector = LogCollector.new "Build contact rollups"
+
+  # Build new daily table of contact rollups
+  ContactRollupsV2.build_contact_rollups(log_collector)
+end
```
Comment (hacodeorg) [inline bin/cron/build_contact_rollups_v2:20]: You can print out content of log_collector by `log_collector.to_s`. We can add error handling and Honeybadger/slack notification later if needed.

**Comment by bencodeorg** [comment]
Separating migration and other logic into separate PRs, closing.

**Pull request merged**